src/crypto/generichash/mod.rs

```diff
@@ -0,0 +1,203 @@
+//! `GenericHash`.
+//!
+#[cfg(not(feature = "std"))] use prelude::Vec;
```

Comment:
Doesn't this break compatibility with embedded systems that do not support heap allocation?

Reply:
Yes, haven't thought of that when I wrote this.
Right now, I can't think of a more ergonomic api which doesn't require `Vec`. Do you have a suggestion?

Reply:
Well, one could provide `State::new` a buffer as a `&mut [u8]` and put that reference in the state instead of a Vec.

Reply:
I think we can put both apis behind `cfg`s, imo that's probably the best way.

Reply:
I think first the `generichash` modules in libsodium_sys should be finished because `crypto_generichash_blake2b_state` and `crypto_generichash_state` are just empty `enums` when they should be structs. The state of BLAKE2b looks like this

```
typedef struct CRYPTO_ALIGN(64) crypto_generichash_blake2b_state {
    uint64_t h[8];
    uint64_t t[2];
    uint64_t f[2];
    uint8_t  buf[2 * 128];
    size_t   buflen;
    uint8_t  last_node;
} crypto_generichash_blake2b_state;
```

in libsodium. You should follow the example of how this is translated to Rust in the SHA2 modules in libsodium-sys. Then I would define `State` in this module simply as a wrapper around the new generic hash state structs you define, following how the hash macros module does it. You don't need to save the `out_len` because this is only used in the call to `init`. Only the `new` method should need this arg passed.

Reply:
Yes I mostly agree. However, I don't see how `out_len` doesn't need to be saved as it's also used to create a ` Digest` (The digest can be variable length but must be `<= crypto_generichash_BYTES_MAX`, compared to the sha2 variant which has a fixed size digest).

Reply:
I bet `out_len` is saved in one of the `crypto_generichash_blake2b_state` fields. If so and you can figure out which one, I would just use that. Otherwise, yes, you're right you have to keep the `out_len` field.

Reply:
Actually, it's not :D I've just checked the internals. This particular function tell us why: 

https://github.com/jedisct1/libsodium/blob/7f7e7235c52f13800df15ef705dbd199252a784c/src/libsodium/crypto_generichash/blake2b/ref/blake2b-ref.c#L161

```c
int blake2b_init_param( blake2b_state *S, const blake2b_param *P )
{
    ...
    p = ( const uint8_t * )( P );
    ...
    S->h[i] ^= LOAD64_LE( p + sizeof( S->h[i] ) * i );
    ...
}
```

So basically, the whole instance of `blake2b_param` is not preserved.

Reply:
I feel like outlen should be saved in libsodium for consistency and ease of use like you're doing here, but that's not really relevant to this thread.